include/adaflo/sharp_interface_util.h

```diff
@@ -803,116 +757,190 @@ collect_evaluation_points(const Triangulation<dim, spacedim> &     surface_mesh,
+        // gather_evaluate normal
+        {
+          buffer_dim.resize(fe_dim.n_dofs_per_cell());
+          for (int i = 0; i < spacedim; ++i)
+            {
+              constraints.get_dof_values(normal_solution.block(i),
+                                         local_dof_indices.begin(),
+                                         buffer.begin(),
+                                         buffer.end());
+              for (unsigned int c = 0; c < cell->get_fe().n_dofs_per_cell(); ++c)
+                buffer_dim[fe_dim.component_to_system_index(i, c)] = buffer[c];
+            }
 
-      // quadrature loop
-      for (unsigned int q = 0; q < n_points; ++q)
+          phi_normal.evaluate(cell,
+                              unit_points,
+                              make_array_view(buffer_dim),
+                              EvaluationFlags::values);
+        }
+
+        // perform operation on quadrature points
+        for (unsigned int q = 0; q < unit_points.size(); ++q)
+          {
+            Assert(phi_normal.get_value(q).norm() > 0, ExcNotImplemented());
+            const auto normal = phi_normal.get_value(q) / phi_normal.get_value(q).norm();
+            phi_force.submit_value(surface_tension * normal * phi_curvature.get_value(q) *
+                                     JxW[q],
+                                   q);
+          }
+
+        // integrate_scatter force
         {
-          Assert(phi_normal.get_value(q).norm() > 0, ExcNotImplemented());
-          const auto normal = phi_normal.get_value(q) / phi_normal.get_value(q).norm();
-          phi_force.submit_value(normal * phi_curvature.get_value(q) * JxW[q], q);
+          buffer_dim.resize(dof_handler_dim.get_fe().n_dofs_per_cell());
+          local_dof_indices.resize(dof_handler_dim.get_fe().n_dofs_per_cell());
+
+          phi_force.integrate(cell, unit_points, buffer_dim, EvaluationFlags::values);
+
+          cell_dim->get_dof_indices(local_dof_indices);
+
+          constraints.distribute_local_to_global(buffer_dim,
+                                                 local_dof_indices,
+                                                 force_vector);
         }
+      }
+  };
 
-      buffer_dim.resize(dof_handler_dim.get_fe().n_dofs_per_cell());
-      local_dof_indices.resize(dof_handler_dim.get_fe().n_dofs_per_cell());
+  std::vector<T> buffer;
 
-      // integrate force
-      phi_force.integrate(cell, unit_points, buffer_dim, EvaluationFlags::values);
+  normal_solution.update_ghost_values();
+  curvature_solution.update_ghost_values();
 
-      cell_dim->get_dof_indices(local_dof_indices);
+  eval.template process_and_evaluate<T>(integration_values, buffer, integration_function);
 
-      constraints.distribute_local_to_global(buffer_dim, local_dof_indices, force_vector);
-    }
+  normal_solution.zero_out_ghost_values();
+  curvature_solution.zero_out_ghost_values();
+  force_vector.compress(VectorOperation::values::add);
```

Comment:
Not that it matters much, but I think the most common notation in deal.II is
```suggestion
  force_vector.compress(VectorOperation::add);
```